scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+		StartTime:   startTime,
+	}
+
+	runs.Runs = append(runs.Runs, TaskRunSimple{RunID: metadata.ID, DateCreated: startTime})
+
+	serializedRuns, err := runs.Marshal()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+
+	serializedMetadata, err := metadata.Marshal()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+	err = m.storage.Create(taskRunKey.String(), string(serializedRuns))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	taskRunMetaKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: startTime}
+	err = m.storage.Create(taskRunMetaKey.String(), string(serializedMetadata))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+
+	return metadata, nil
+}
+
+func (m *TaskMetadataManager) GetRunByID(taskID TaskID, runID TaskRunID) (TaskRunMetadata, fferr.GRPCError) {
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskRunMetadata, err := m.storage.Get(taskRunKey.String())
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	runs := TaskRuns{}
+	err = runs.Unmarshal([]byte(taskRunMetadata))
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	// Want to move this logic out
+	found, runRecord := runs.ContainsRun(runID)
+	if !found {
+		err := fferr.NewKeyNotFoundError(taskRunKey.String(), fmt.Errorf("run not found"))
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	date := runRecord.DateCreated
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: runRecord.RunID, date: date}
+	rec, err := m.storage.Get(taskRunMetadataKey.String())
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	taskRun := TaskRunMetadata{}
+	err = taskRun.Unmarshal([]byte(rec))
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+	return taskRun, nil
+}
+
+func (m *TaskMetadataManager) GetRunsByDate(start time.Time, end time.Time) (TaskRunList, fferr.GRPCError) {
+	// the date range is inclusive
+	var runs []TaskRunMetadata
+
+	// iterate through each day in the date range including the end date
+	for date := start; date.Before(end) || date.Equal(end); date = date.AddDate(0, 0, 1) {
+		dayRuns, err := m.getRunsForDay(date, start, end)
```

Comment:
Why do we do it like this?

Reply:
assuming we need a "day" function I image the right logic would be that the day uses the get runs by date and just truncates the times. Either way the day function confuses me a bit

Reply:
This will be used to display the Runs on the dashboard. It allows a filtering based on date range.